// jovo-platform-alexa/dist/src/modules/Display.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _get = require("lodash.get");
const _set = require("lodash.set");
const _merge = require("lodash.merge");
const jovo_core_1 = require("jovo-core");
const AlexaSkill_1 = require("../core/AlexaSkill");
class Display {
    install(alexa) {
        alexa.middleware('$type').use(this.type.bind(this));
        alexa.middleware('$output').use(this.output.bind(this));
        /**
         * Shows template on Echo Show
         * @public
         * @param {*} template
         * @return {AlexaSkill}
         */
        AlexaSkill_1.AlexaSkill.prototype.showDisplayTemplate = function (template) {
            _set(this.$output, 'Alexa.DisplayTemplate', new DisplayRenderTemplateDirective(template));
            return this;
        };
        /**
         * Shows hint on Echo Show
         * @public
         * @param {*} text
         * @return {AlexaSkill}
         */
        AlexaSkill_1.AlexaSkill.prototype.showHint = function (text) {
            _set(this.$output, 'Alexa.DisplayHint', new DisplayHintDirective(text));
            return this;
        };
        /**
         * Adds apl directive
         * @deprecated Please use addAPLDirective()
         * @public
         * @param {*} directive
         * @return {AlexaSkill}
         */
        AlexaSkill_1.AlexaSkill.prototype.addAplDirective = function (directive) {
            const directives = _get(this.$output, 'Alexa.Apl', []);
            directives.push(directive);
            _set(this.$output, 'Alexa.Apl', directives);
            return this;
        };
        /**
         * Adds apl directive
         * @public
         * @param {*} directive
         * @return {AlexaSkill}
         */
        AlexaSkill_1.AlexaSkill.prototype.addAPLDirective = function (directive) {
            const directives = _get(this.$output, 'Alexa.Apl', []);
            directives.push(directive);
            _set(this.$output, 'Alexa.Apl', directives);
            return this;
        };
        /**
         * Adds apl directive
         * @public
         * @param {*} documentDirective
         * @return {AlexaSkill}
         */
        AlexaSkill_1.AlexaSkill.prototype.addAPLDocument = function (documentDirective) {
            const document = {
                type: 'Alexa.Presentation.APL.RenderDocument',
                version: '1.0',
            };
            _merge(document, documentDirective);
            const directives = _get(this.$output, 'Alexa.Apl', []);
            directives.push(document);
            _set(this.$output, 'Alexa.Apl', directives);
            return this;
        };
        /**
         * Adds apl directive
         * @public
         * @param {string} token
         * @param {*} commands
         * @return {AlexaSkill}
         */
        AlexaSkill_1.AlexaSkill.prototype.addAPLCommands = function (token, commands) {
            const commandDirective = {
                type: 'Alexa.Presentation.APL.ExecuteCommands',
                token,
                commands,
            };
            const existingExecuteCommands = _get(this.$output, 'Alexa.Apl', []).filter((directive) => {
                return directive.type === commandDirective.type && directive.token === commandDirective.token;
            });
            if (existingExecuteCommands[0]) {
                existingExecuteCommands[0].commands = existingExecuteCommands[0].commands.concat(commands);
            }
            else {
                const directives = _get(this.$output, 'Alexa.Apl', []);
                directives.push(commandDirective);
                _set(this.$output, 'Alexa.Apl', directives);
            }
            return this;
        };
        /**
         * Shows video on Echo Show
         * @public
         * @param {string} url
         * @param {string} title
         * @param {string} subtitle
         */
        AlexaSkill_1.AlexaSkill.prototype.showVideo = function (url, title, subtitle) {
            _set(this.$output, 'Alexa.VideoApp', (new VideoAppLaunchDirective()).setData(url, title, subtitle));
            return this;
        };
    }
    uninstall(alexa) {
    }
    type(alexaSkill) {
        const alexaRequest = alexaSkill.$request;
        if (_get(alexaRequest, 'request.type') === 'Display.ElementSelected' ||
            _get(alexaRequest, 'request.type') === 'Alexa.Presentation.APL.UserEvent') {
            alexaSkill.$type = {
                type: jovo_core_1.EnumRequestType.ON_ELEMENT_SELECTED,
                subType: _get(alexaRequest, 'request.token')
            };
        }
    }
    output(alexaSkill) {
        const output = alexaSkill.$output;
        const response = alexaSkill.$response;
        if (alexaSkill.$request.hasScreenInterface()) {
            if (_get(output, 'Alexa.DisplayTemplate')) {
                const directives = _get(alexaSkill.$response, 'response.directives', []);
                directives.push(_get(output, 'Alexa.DisplayTemplate'));
                _set(response, 'response.directives', directives);
            }
            if (_get(output, 'Alexa.DisplayHint')) {
                const directives = _get(alexaSkill.$response, 'response.directives', []);
                directives.push(_get(output, 'Alexa.DisplayHint'));
                _set(response, 'response.directives', directives);
            }
            if (_get(output, 'Alexa.VideoApp')) {
                const directives = _get(alexaSkill.$response, 'response.directives', []);
                directives.push(_get(output, 'Alexa.VideoApp'));
                _set(response, 'response.directives', directives);
                if (_get(response, 'response.shouldEndSession')) {
                    delete response.response.shouldEndSession;
                }
                // set sessionAttributes (necessary since AlexaCore's handler runs before us
                // and skips adding session attributes due to shouldEndSession being true at that point)
                if (alexaSkill.$session && alexaSkill.$session.$data) {
                    _set(response, 'sessionAttributes', alexaSkill.$session.$data);
                }
            }
        }
        if (alexaSkill.$request.hasAPLInterface()) {
            if (_get(output, 'Alexa.Apl')) {
                let directives = _get(response, 'response.directives', []);
                if (Array.isArray(_get(output, 'Alexa.Apl'))) {
                    directives = directives.concat(_get(output, 'Alexa.Apl'));
                }
                else {
                    directives.push(_get(output, 'Alexa.Apl'));
                }
                _set(response, 'response.directives', directives);
            }
        }
        if (_get(output, 'Alexa.deleteShouldEndSession')) {
            if (_get(response, 'response.shouldEndSession')) {
                delete response.response.shouldEndSession;
            }
        }
    }
}
exports.Display = Display;
class DisplayDirective {
    constructor(type) {
        this.type = type;
    }
}
class DisplayRenderTemplateDirective extends DisplayDirective {
    constructor(template) {
        super('Display.RenderTemplate');
        this.template = template;
    }
    setTemplate(template) {
        this.template = template;
        return this;
    }
}
class DisplayHintDirective extends DisplayDirective {
    constructor(text) {
        super('Hint');
        if (text) {
            this.setHint(text);
        }
    }
    setHint(text) {
        this.hint = {
            type: 'PlainText',
            text,
        };
        return this;
    }
}
class VideoAppLaunchDirective extends DisplayDirective {
    constructor() {
        super('VideoApp.Launch');
    }
    setVideoItem(videoItem) {
        this.videoItem = videoItem;
    }
    setData(url, title, subtitle) {
        this.videoItem = {
            source: url,
        };
        if (title) {
            this.videoItem.metadata = {
                title
            };
            if (subtitle) {
                this.videoItem.metadata.subtitle = subtitle;
            }
        }
        return this;
    }
}
